### DISCUSSION TITLE
axum-extra - cookie not being updated up on response but I can see it in the headers

### DISCUSSION THREAD

**Discussion by yatesco** [tokio-rs/axum#1223] [Q&A] [answered]
Hi there,

I'm using `private-cookie` from `axum-extra` and following the example at https://docs.rs/axum-extra/latest/axum_extra/extract/struct.PrivateCookieJar.html. I can see (in Chrome) the request and response contain the cookie/set-cookie headers, but any updates to the cookie aren't persisted on the next request.

1. request 1 - cookie is empty so a default is set and returned
2. request 2 - the client POSTS an update
3. server picks up the existing cookie, updates it and returns it
4. request 3 - the client POSTS another update
5. server opens the jar but the state doesn't include the update in step 3 - it is still the default value

I have no other extractors, but I do have another `Extension` which adds in some other state.

Relevant code:
```rust

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    // build our application with some routes
    // let app = Router::new().route("/greet/:name", get(greet));
    let app = Router::new()
        .route("/", get(root))
        .route("/expand/:id", post(toggle_expand))
        .layer(extract::Extension(reference_data))
        .layer(Extension(key));
...
}

async fn root(
    jar: PrivateCookieJar,
    Extension(rd): Extension<Arc<RwLock<reference_data_domain::ReferenceData>>>,
) -> impl IntoResponse {
    let state = get_state(&jar);
    render_root(state, jar, rd.clone())
}

async fn toggle_expand(
    jar: PrivateCookieJar,
    Path(id): Path<usize>,
    Extension(rd): Extension<Arc<RwLock<reference_data_domain::ReferenceData>>>,
) -> impl IntoResponse {
    let mut state = get_state(&jar);
   // *** MAKE A CHANGE TO THE STATE
    render_root(state, jar, rd.clone())
}

fn render_root(
    state: api::State, 
    jar: PrivateCookieJar,
    rd: Arc<RwLock<reference_data_domain::ReferenceData>>,
) -> (PrivateCookieJar, maud::Markup) {
    dbg!(&state); // *** THIS SHOWS THE UPDATED STATE ***/
    let jar = set_state(jar, &state);
    let markup = html! {
        (DOCTYPE)
            meta charset="utf-8";
          ....
        }
    };
    (jar, markup)
}

fn get_state(jar: &PrivateCookieJar) -> api::State {
    let value: Option<String> = jar
        .get("qfi-pj-ui")
        .map(|cookie| cookie.value().to_owned());
    println!("cookie value: {:?}", &value); // *** FIRST TIME IS EMPTY BUT THEN ALWAYS SHOWS DEFAULT VALUE ***
    let state = match value {
        Some(json) => serde_json::from_str(&json).expect("cannot deserialise from cookie"),
        None => api::State::new(})
    };
    state
}

fn set_state(jar: PrivateCookieJar, state: &api::State) -> PrivateCookieJar {
    let persistent_state = serde_json::to_string(&state).expect("cannot serialise model");
    println!("persisting state: {}", persistent_state); /*** THIS CORRECTLY SHOWS THE UPDATED STATE ***/
    let jar = jar.add(Cookie::new("qfi-pj-ui",persistent_state));
    jar
}
```

Chrome example of a response:
<img width="640" alt="image" src="https://user-images.githubusercontent.com/181469/183050191-7fe23a43-bee1-462c-a204-16d03c70fb06.png">

**Comment by jplatte** [comment, marked as answer]
How does this post request happen? Is it via navigation, i.e. submitting a form, or is JavaScript involved? Also, is there a small info sign on the right of the set-cookie header in the dev console? The screenshot is cut off on the right. 

edit: Conclusion of the discussion below is that the provided app sets multiple values for the same cookie, but with different paths. How the code could be changed to actually be able to update the cookie value is yet unclear, but it's clear that both `set-cookie` headers are being received and respected by the browser.

**Reply by yatesco** [reply]
thanks for replying - it uses htmx.org to do a POST and no, there isn't a small info sign.

**Reply by jplatte** [reply]
Right, so it's AJAX. But since this isn't cross-origin, I think it should just work 🤔

The small info sign should also show up when a set-cookie is ignored for some reason. Maybe try in Firefox to see whether it has the same behavior / shows more info somewhere?

**Reply by yatesco** [reply]
one thing that is maybe interesting - there are multiple cookies with the same name in Chrome:
first request:
<img width="1023" alt="image" src="https://user-images.githubusercontent.com/181469/183055908-fde12e17-d213-4a2c-b0ba-ea6168c63436.png">

second request:
<img width="1025" alt="image" src="https://user-images.githubusercontent.com/181469/183055860-f597503d-1953-426e-81b5-4c6cc309a92f.png">

and the Cookie that gets sent through to the server is *always* the one associated with "/", not the one returned from the "/expand/12345"....:
```rust
[proposed-destination/src/main.rs:131] jar = PrivateCookieJar {
    jar: CookieJar {
        original_cookies: {
            DeltaCookie {
                cookie: Cookie {
                    cookie_string: Some(
                        " qfi-pj-ui=s+1Mae%2F8eROGgWOv0+JLQamYTsjPOSpOxXp5QPRIFrmWXl+99MihUIYmTpOw%2FVzyBC236+r8BUcw3DQN85dMtdtE7xNbwC%2FE32f36+TFxB9V2s8WEGry0SQidznT9U3LcOFWzjFj",
                    ),
                    name: Concrete(
                        "qfi-pj-ui",
                    ),
                    value: Concrete(
                        "YW0VC7A8EYPWF9hpQ5vqQsHs//cHrVyJGngsI4wCLIGfrlyb4Pg0IyPhRWQ3u3Mqfx+OIWnWrOmjh1+Lf3YsZiDWAPMqXlU8+gEN8atqvgJU1of8KJPUDrBGPmsOxU7uebcK4E1/",
                    ),
                    expires: None,
                    max_age: None,
                    domain: None,
                    path: None,
                    secure: None,
                    http_only: None,
                    same_site: None,
                },
                removed: false,
            },
        },
        delta_cookies: {},
    },
    key: "REDACTED",
}
```

So it seems that the cookie that is set at "/" takes precedence over the cookie that is set at "/any/path"?

**Reply by jplatte** [reply]
What I also find interesting is that it looks like the *request* headers in your snippet contain multiple values for this cookie? Maybe that's related to your issue.

**Reply by yatesco** [reply]
Yes - I somehow need to say "this cookie from /nested/path replaces the cookie from /". I naively thought cookies were scoped to a domain, not a path. TIL.

**Reply by jplatte** [reply]
No, both are sent, that's what Chromium shows. It might just be a limitation in the cookie crate that it (seemingly) discards the extra occurrence of the same cookie.

**Reply by jplatte** [reply]
Try setting the path of the cookie explicitly, like in the second example in cookie's docs: https://docs.rs/cookie/latest/cookie/struct.Cookie.html#constructing-a-cookie

**Reply by yatesco** [reply]
I suspect I need more access to the actual header itself, to define the path the cookie is associated with... I'll keep playing - thanks for your help :-)

**Reply by jplatte** [reply]
In any case it seems pretty clear that if there is a bug on the server side here, it's the `cookie` library not allowing you to access multiple values for the same cookie. I'll mark this as answered.

**Comment by yatesco** [comment]
For anyone keeping track - POSTing to "/" works fine, it is setting the cookie on requests from "/" and "/nested/path" breaks as it sends both cookies back to the server and the one associated with "/" is the one returned by `PrivateCookieJar::get`.
